debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java

```diff
@@ -41,6 +55,33 @@ public Connector getConnector() {
         return new MongoDbConnector();
     }
 
+    private ConnectionString enforceReadPreferencePrimary(ConnectionString connectionString) {
+        URI connectionStringUrl = UriBuilder.fromUri(connectionString.getConnectionString())
+                .replaceQueryParam("readPreference", "primary")
```

Comment:
There are several users who were very explicit(e.g. Okta) about having a control over what nodes debezium interacts with. Selecting collection is not crucial to the point where we need the primary node -- If the rest of the connector itself doesn't need it, UI should not be the component mandating this. 


@jpechane @Naros  what's your take on this?

Reply:
The primary node in a low consistency (or tunebale consistency) datastore like Mongo DB is the only reliable source for metadata like active tables. This was used explicitly for this reason and especially bigger, busy cluster could return unexpected metadata from my experiences.

Reply:
Rene... I spend literary hundreds of hours working with the mongo code. Including all the comments and Jiras we historically had. I'm pretty sure that I know why things were done one way or the other in the past. I also interact with our users regarding this on daily basis and especially in busy environments they don't like their primary getting touched! 

As for the "freshness" of metadata -- I already acknowledged that you are **technically** correct. However my argument is that for the purpose of UI we don't need that level of "freshness". IMHO the case of somebody going to create a collection and then immediately jumping into Debezium UI and looking for it the for the purpose of setting up a connector is highly unlikely.

Reply:
Wouldn't it also make sense in this context to rely on the data in the `secondary` if that's what they are connecting with rather than the primary to avoid inconsistencies? I could imagine deploying a connector and pointing to the `secondary` and realizing the collections that are of interest not being visible could be considered a warning or light bulb for the user to check their replication status, no?

In the other case, fetching this detail from the primary and deploying the connector could give the user a false sense of everything being okay. Then, when the connector is running and not capturing changes, this could be seen more as a Debezium problem rather than checking their environment, no?

Given this example, using the secondary in this case makes logical sense, as it can help the user identify potential issues before the connector is even deployed.

Reply:
Personally I think we are getting into details which are not really important. In practice the users is IMHO much more likely to be configuring the connector to capture data from collection which does not exist yet (e.g. user wants to capture changes to all "department" collection from inventory database. `collection.include.list=inventory\.department_.*`) rather than facing an issue where the delay between primary and secondary is so large that it is not yet visible in the secondary (a problem which is easily solvable by a presence of "refresh" button next to the collection list).

Reply:
> Wouldn't it also make sense in this context to rely on the data in the secondary if that's what they are connecting with rather than the primary to avoid inconsistencies? I could imagine deploying a connector and pointing to the secondary and realizing the collections that are of interest not being visible could be considered a warning or light bulb for the user to check their replication status, no?

I see it from the other side. The user always sees the correct collections all the time when using the correct list of collections from the primary and eventually the system (replica node/s) comes into a consistent state and the initially selected collections are streamed from the secondary. So the right set of collections is the one on the primary not the info from the  replica nodes. When using the info from a random node the user might see collections that got removed on the primary but are not yet replicated to the replica node that the user connects to. After a certain amount of time the system will be in the state of the primary, not in the state of one of the replica nodes.
Btw the default ReadPreference mode should be `primary` and that means per default the client should connect to the primary anyways, no?

Furthermore, the UI is not hammering that endpoint. Actually the user needs to configure the filter config options and then press a button to make the request to validate the inputs and get the list of collections returned. I don't think the argument about 
The only attack vector is that someone is hammering this endpoint on purpose but this would be an attack on purpose and can be handled by adding a circuit breaker in front of the API.

Reply:
The default is whatever the standard mongo default is. Which is indeed `primary`. I have no problem with reaching to primary if the user doesn't specify the read preference or explicitly sets it to `primary` :).

Reply:
@jpechane  we obviously have a disagreement on this topic. What would be your preferred way of handling this?

Reply:
@rk3rn3r I would keep this unresolved until we have Jiri's take on this. If multiple team members think we should do it this way then I'm resting my case ;).

Reply:
I am not going to change the code in a way that will lead to wrong results sooner or later.
From my best current knowledge replica nodes can have the wrong view about the active collections and other cluster metadata and thus might return results that are already outdated or might be outdated eventually.
If you can prove that those metadata are consistent I will surely change this now but I don't see any good reason to change this for the fore-mentioned reasons.
For example your point that this might cause stability issues of the primary node: like I mentioned before, those requests are only happening when the user validates the filter list and even with a very high number of collections I am not convinced this will harm the primary's stability. I am very sorry and although I see your point I just can't do this from my developer's ethical attitude.

What I could imagine is to offer an option to relax this behavior: I can file a Jira task to add a `filter.validation.from.replicas` config option that defaults to `false`. When enabled by setting it to `true` it will change the behavior to use replica nodes. If you don't find this compromise acceptable, feel free to file a Jira for your preferred solution and change it yourself after this PR but I am not putting my name on lines of code that I can't prove to be correct. Honestly: sorry again!

Reply:
Okay, thanks for that comment. I think your points are valid concerns. Such scenario would all require to set `directConnection` option in the connection string, right? That means in the case that `directConnection` is in the connection string I should/must not do "adjustments" to the connection string. But in other cases when there is no `directConnection` parameter set it should be okay? Did I get this right?

Reply:
@rk3rn3r not the one with tags (that's essentially just scoping the set of usable nodes for this connection string). Relaxing to `primaryPreferred` would work for that case. 

Which once again demonstrates my point about hidden complexity and "smart adjustments".

Reply:
I really don't think there should be any adujstment/optimizaion to connection setup in the REST extension. If something is broken it has to be fixed in the connector/core code. If not then REST extension should not introduce any change in default behaviour.

Reply:
@jpechane  just to be clear. This doesn't change the configuration for the connector. It's a temporary change to the connection string just for the purpose of the connection used by the REST extension.

Reply:
Yes, I understand. My take is that we ould use the same connection setup when checking the filters for UI usecase like we do for connector. As soon as we'd allow differences we might end up in strange discrepances that would be hard to debug in future.

Reply:
> As soon as we'd allow differences we might end up in strange discrepances that would be hard to debug in future.

I'm sorry @jpechane , but I can't agree on this one. A cluster will eventually reach the same status as the primary node and not the other way around (the state of a replica node). That means if you are not reading the fresh metadata of the primary node then you will end up facing such kind of discrepancies that you mentioned, for example: why was this topic created (by Debezium) or shown in the UI but no data are coming anymore? Answer: the replica node is not aware of the removal of that collection on the primary node (because it lags behind) but still returns it in the list of collections.
Speaking generally for the connector I would even assume that metadata should be read fresh from the primary even when the replication happens from a replica/secondary node. And it should furthermore ignore for example collections that were removed on the primary node but are still available on that replica node.

Reply:
@rk3rn3r  what metadata would that be in our specific implementation? In a theoretical scenario where collection doesn't yet appear to be present during snapshot the event's will be streamed anyway later. 

Additionally all you are saying is extremely time dependant. I don't think complexity is worth it if the situation would be different, had the user decide to deploy the connector a minute later. Again, this is not the hot path.

Reply:
BTW how do we deal with this in case of relational databases? Things like PG replicas. Oracle Stand-By and so on? I know very little about these but I think the situation might be similar on that front.

Reply:
I think we're all caught up in the data technicalities but we don't seem to focus on what I believe are more critical issues: trust, security, and expectations.

If I configure a tool to connect to a replica/secondary server and then it blatantly disregards that configuration and connects to my primary, that can be dangerous for any of those aforementioned reasons.  It's one thing if I don't specify a preference, there aren't any rules pre-established for which we use, and we connect to the primary; however, if we've established those rules, those should be followed -- regardless of the eventual consistency concerns anyone may have.  

If we begin to change the strategy in the UI that doesn't align with the connector's strategy, then you've eroded trust, you've made it even more confusing for the user, especially if they've expressly specified to use the secondary in the connection string, and could lead to a DBA putting "can't deploy to prod" order in place for new tooling because they strictly forbid primary connections -- which is more common than you might think in overly exacerbated environments where they're already working at near max capacity.

There is nothing saying we can't put a small indicator in the UI that alerts the user that we may be connecting to a secondary and that the list of collections/tables shown may not be absolutely accurate due to using the replica rather than the primary -- that solves @rk3rn3r's concerns and it allows the process to be consistent with the connector's behavior without an entirely new set of code around connection management.

Reply:
@jpechane I filed https://issues.redhat.com/browse/DBZ-7165 for a suggestion task to change the behavior of the connector towards that higher consitency model that we recently gave up. imo we were always reading from the primary before, no?

I filed https://issues.redhat.com/browse/DBZ-7166 for aligning the REST extension behavior with whatever we decide as outcome of DBZ-7165 above. In order to align the behavior of the REST endpoint and the connector.

I also filed https://issues.redhat.com/browse/DBZ-7167 to move the filtering logic from the "outside" in the REST extension code into the "inside" of the connector implementations / the connector classes.

Reply:
Done